src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs

```diff
@@ -2,35 +2,96 @@
+
         protected IStreamQueueCheckpointer<string> Checkpointer { private set; get; }
 
-        protected EventHubQueueCache(IStreamQueueCheckpointer<string> checkpointer, ICacheDataAdapter<EventData, TCachedMessage> cacheDataAdapter, ICacheDataComparer<TCachedMessage> comparer)
-            : base(cacheDataAdapter, comparer)
+        protected EventHubQueueCache(int defaultMaxAddCount, IStreamQueueCheckpointer<string> checkpointer, ICacheDataAdapter<EventData, TCachedMessage> cacheDataAdapter, ICacheDataComparer<TCachedMessage> comparer)
         {
-            cacheDataAdapter.PurgeAction = Purge;
+            this.defaultMaxAddCount = defaultMaxAddCount;
             Checkpointer = checkpointer;
-            OnPurged = CheckpointOnPurged;
+
+            cache = new PooledQueueCache<EventData, TCachedMessage>(cacheDataAdapter, comparer);
+            cacheDataAdapter.PurgeAction = cache.Purge;
+            cache.OnPurged = CheckpointOnPurged;
+
+            cachePressureMonitor = new CachePressureMonitor();
         }
 
+        protected abstract string GetOffset(TCachedMessage lastItemPurged);
+
+        // cachePressureContribution should be a double between 0-1, indicating how much danger the item is of being removed from the cache.  0 indicating  no danger, 1 indecating removal is imminent.
+        protected abstract bool TryCalculateCachePressureContribution(StreamSequenceToken token, out double cachePressureContribution);
+
         public void Dispose()
         {
-            OnPurged = null;
+            cache.OnPurged = null;
         }
 
-        protected abstract string GetOffset(TCachedMessage lastItemPurged);
-
         private void CheckpointOnPurged(TCachedMessage lastItemPurged)
         {
             Checkpointer.Update(GetOffset(lastItemPurged), DateTime.UtcNow);
         }
+
+        public int MaxAddCount { get { return cachePressureMonitor.IsUnderPressure() ? 0 : defaultMaxAddCount; } }
+
+        public void Add(EventData message)
+        {
+            cache.Add(message);
+        }
+
+        public object GetCursor(IStreamIdentity streamIdentity, StreamSequenceToken sequenceToken)
+        {
+            return cache.GetCursor(streamIdentity, sequenceToken);
+        }
+
+        public bool TryGetNextMessage(object cursorObj, out IBatchContainer message)
+        {
+            if (!cache.TryGetNextMessage(cursorObj, out message))
+                return false;
+            double cachePressureContribution;
+            if (TryCalculateCachePressureContribution(message.SequenceToken, out cachePressureContribution))
+            {
+                cachePressureMonitor.RecordCachePressureContribution(cachePressureContribution);
+            }
+            return true;
+        }
+
+        private class CachePressureMonitor
+        {
+            const double pressureThreshold = 1.0/3.0;
+
+            private double accumulatedCachePressure;
+            private int cachePressureContributionCount;
```

Comment:
But again, not saying you must add the additional capabilities in this PR, just saying that we should acknowledge the fact that there won't be a single default that everyone likes, so we might as well name this implementation appropriately and not as if it will be the single one that needs no qualification.